### PR TITLE
add decomposition for frexp

### PR DESCRIPTION
**Pull request by isuruf**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120429
* __->__ #119217
* #120027
* #119284

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames

### CODE REVIEW

File: torch/_inductor/codegen/triton.py [hunk 1]
```diff
@@ -660,6 +660,14 @@ def atan(x):
     def atanh(x):
         return f"tl.math.atanh({x})"
 
+    @staticmethod
+    def frexp0(x):
+        return f"tl.math.ilogb({x})"
+
+    @staticmethod
+    def frexp1(x):
+        return f"tl.math.ldexp({x}, -tl.math.ilogb({x}))"
```
Comment (peterbell10) [inline torch/_inductor/codegen/triton.py:669]: Why not make the operators `ilogb` and `ldexp`?

File: torch/_inductor/lowering.py [hunk 2]
```diff
@@ -5179,6 +5202,8 @@ def register_pointwise_numeric_ldf64(op):
 register_pointwise_numeric(aten.log10)
 register_pointwise_numeric(aten.nextafter)
 
+register_frexp()
```
Comment (peterbell10) [inline torch/_inductor/lowering.py:5205]: Not sure I see this point in splitting this into two places?

File: torch/_decomp/__init__.py [hunk 3]
```diff
@@ -298,6 +298,7 @@ def core_aten_decompositions() -> Dict[torch._ops.OperatorBase, Callable]:
             aten.floor_divide,
             aten.frac,
             aten.frac_,
+            aten.frexp,
```
Comment (peterbell10) [inline torch/_decomp/__init__.py:301]: `core_aten_decompositions` should not decompose into prims. We should just lower `aten.frexp` directly in inductor.

File: torch/_inductor/codegen/triton.py (hunk 1 above)
Comment (isuruf) [reply torch/_inductor/codegen/triton.py:669]: Doing `ilogb` + `ldexp` to get `frexp` is a bit slower than just doing `frexp` and using `frexp` directly for the cpu device.

File: torch/_inductor/decomposition.py [hunk 4]
```diff
@@ -43,6 +43,7 @@
         aten.dist,
         aten.empty_like,
         aten.flip,
+        aten.frexp,
```
Comment (peterbell10) [inline torch/_inductor/decomposition.py:46]: We don't need this decomposition for inductor either. Note that the decomposition will be used  as the "meta"-implementation, which does shape and dtype propagation when tracing. So it wasn't a waste of time.

File: torch/_refs/__init__.py [hunk 5]
```diff
@@ -1366,6 +1366,11 @@ def fmod(a: TensorLikeType, b: TensorLikeType) -> TensorLikeType:
     return prims.fmod(a, b)
 
 
+@register_decomposition(aten.frexp)
+def frexp(self: TensorLikeType) -> Tuple[TensorLikeType, TensorLikeType]:
```
Comment (peterbell10) [inline torch/_refs/__init__.py:1370]: This should have a `PythonRefInfo` in `torch/testing/_internal/common_methods_invocations.py`. Also, `frexp` does have an out overload so `@out_wrapper` would be appropriate.

File: torch/_inductor/codegen/triton.py (hunk 1 above)
Comment (peterbell10) [reply torch/_inductor/codegen/triton.py]: Hrm looks like these functions aren't inlined in C++ but the compiler is able to cse the two frexp calls. Would be nicer if we could CSE at inductor level but I guess this is okay.

Comment (isuruf) [reply torch/_inductor/codegen/triton.py]: Yes, I noticed that too. The compiler is however able to optimize the two calls into one.

File: torch/_refs/__init__.py (hunk 5 above)
Comment (isuruf) [reply torch/_refs/__init__.py:1370]: For the out wrapper, I'm not sure how to do that. `torch.frexp` has a `out` parameter which takes in a tuple of Tensors and `aten.frexp` has two out parameters `mantissa, exponent`. I'm not sure how to reconcile this difference.
Setting `@out_wrapper("mantissa", "exponent")` fails in an error.

The other methods that have different names like `std_mean, var_mean` have `supports_out=False` in their opinfos.

Comment (isuruf) [reply torch/_refs/__init__.py:1371]: Thanks. That fixed the issues for me.

File: torch/_inductor/virtualized.py [hunk 6]
```diff
@@ -497,6 +497,9 @@ def erfc(self, x0: T) -> T:
     def erfinv(self, x0: T) -> T:
         ...
 
+    def frexp(self, x0: T) -> Tuple[T, T]:
```
Comment (peterbell10) [inline torch/_inductor/virtualized.py:500]: This is wrong.

**Review by peterbell10 — CHANGES_REQUESTED** [review]
Mostly just removing xfails left to do.

File: torch/_inductor/virtualized.py [hunk 7]
```diff
@@ -829,6 +827,13 @@ def __mod__(self, other):
     def __pow__(self, other):
         return ops.pow(self, other)
 
+    def __getitem__(self, items):
```
Comment (peterbell10) [inline torch/_inductor/virtualized.py:830]: This should never happen and indicates something is wrong elsewhere. `__getitem__` should be being called on an actual tuple, and that tuple shouldn't have been wrapped in an `OpsValue` wrapper.

File: torch/_inductor/codegen/cpp.py [hunk 8]
```diff
@@ -825,23 +825,14 @@ def copysign(x, y):
...
-
-    @staticmethod
-    def frexp1(x):
-        code = BracesBuffer()
-        exponent = V.kernel.cse.newvar()
-        code.writeline(f"int32_t {exponent};")
-        code.writeline(f"std::frexp({x}, &{exponent});")
-        V.kernel.compute.splice(code)
-        return exponent
+        return mantissa, exponent
```
Comment (peterbell10) [inline torch/_inductor/codegen/cpp.py:835]: This still won't actually get CSE'd because `writeline` writes directly to the output buffer, bypassing the CSE mechanism. To get this to CSE you would need to do the caching manually.

Something like:
```python
cache_key = f"frexp({x})"
if cache_key in V.kernel.compute.cache:
    return V.kernel.compute.cache[cache_key]

# ... generate frexp call ...

V.kernel.compute.cache[cache_key] = (mantissa, exponent)
return mantissa, exponent
```

And this should really be defined as a method on the kernel itself since we're doing direct buffer access.

File: torch/_inductor/codegen/common.py [hunk 9]
```diff
@@ -1127,13 +1132,26 @@ def inner(*args, **kwargs):
...
+                            result.append(csevar)
+                        return result
+                    else:
+                        csevar = self.cse.generate(
+                            self.compute,
+                            value,
+                            bounds=buf_bounds,
+                        )
+                        csevar.update_on_args(name, args, kwargs)
+                        return csevar
```
Comment (peterbell10) [inline torch/_inductor/codegen/common.py:1154]: How about:

```suggestion
                    def do_cse(v):
                        csevar = self.cse.generate(...)
                        csevar.update_on_args(...)
                        return csevar

                    return pytree.tree_map(do_cse, value)
```

File: torch/_inductor/codegen/cpp.py [hunk 10]
```diff
@@ -2640,7 +2634,12 @@ def inner(*args, **kwargs):
 
                     if name not in self.fast_vec_list:
                         self.disable_vec(f"op: {name}")
-                    return self.simd_vec
+
+                    parent_val = getattr(parent_handler, name)(*args, **kwargs)
+                    try:
+                        return tuple([self.simd_vec] * len(parent_val))
+                    except TypeError:
+                        return self.simd_vec
```
Comment (peterbell10) [inline torch/_inductor/codegen/cpp.py:2642]: Would this work?
```suggestion
                    return pytree.tree_map(lambda _: self.simd_vec, parent_val)
```

**Review by peterbell10 — CHANGES_REQUESTED** [review]

**Comment by isuruf** [comment]
This is ready for a review. Note that the failure is seen on https://github.com/pytorch/pytorch/pull/119227 too

File: torch/_inductor/ops_handler.py [hunk 11]
```diff
@@ -624,13 +636,17 @@ def inner(*args, **kwargs):
...
-                self.op_count += 1
-                self.var_names[val] = varname
-                return varname
-            else:
-                return self.var_names[val]
+
+            def count(val):
+                if val not in self.var_names:
+                    varname = f"tmp{self.op_count}"
+                    self.op_count += 1
```
Comment (peterbell10) [inline torch/_inductor/ops_handler.py:643]: A bit weird that frexp counts as two ops but okay.

File: torch/_subclasses/fake_tensor.py [hunk 12]
```diff
@@ -166,8 +166,7 @@ def torch_decomp_decompositions(func):
     from torch._decomp import decomposition_table
 
     decompositions = torch._decomp.decompositions
-    decomp_attrs = [getattr(decompositions, attr) for attr in dir(decompositions)]
-    return decomposition_table[func] in decomp_attrs
+    return decomposition_table[func].__name__ in dir(decompositions)
```
Comment (peterbell10) [inline torch/_subclasses/fake_tensor.py:169]: What's the issue here?

Comment (isuruf) [reply torch/_subclasses/fake_tensor.py:169]: Previously `register_decomposition` was changing the reference functions to allow for differences in `out` handling in torch public API vs aten API. We fixed the issue by only changing the function in the decomposition table and keeping the original function where it was defined in `torch/_decomp/decompositions.py`.

Since the function in the decomposition table is different to the function that is referred in `torch._decomp.decompositions`, this function thinks they are different. This change fixes that.

File: test/test_unary_ufuncs.py [hunk 13]
```diff
@@ -385,9 +391,15 @@ def test_non_contig_expand(self, device, dtype, op):
             contig = op(contig, **torch_kwargs)
             non_contig = op(non_contig, **torch_kwargs)
             for i in range(3):
-                self.assertEqual(
-                    contig, non_contig[i], msg="non-contiguous expand[" + str(i) + "]"
-                )
+                if is_iterable_of_tensors(contig):
+                    for a, b in zip(contig, non_contig):
+                        self.assertEqual(
+                            a, b[i], msg="non-contiguous expand[" + str(i) + "]"
```
[the remark this answers is not in the document]
Comment (isuruf) [reply test/test_unary_ufuncs.py]: Yes, we can. Fixed it now

File: test/test_unary_ufuncs.py [hunk 14]
```diff
@@ -433,7 +445,13 @@ def test_batch_vs_slicing(self, device, dtype, op):
 
         torch_kwargs, _ = op.sample_kwargs(device, dtype, input)
         actual = op(input, **torch_kwargs)
-        expected = torch.stack([op(slice, **torch_kwargs) for slice in input])
+
+        expected_unstacked = [op(slice, **torch_kwargs) for slice in input]
```
[the remark this answers is not in the document]
Comment (isuruf) [reply test/test_unary_ufuncs.py]: Removed it. Is there a way to use tree_map here as well?

File: torch/_inductor/ops_handler.py (hunk 11 above)
Comment (isuruf) [reply torch/_inductor/ops_handler.py:643]: Agreed. Not sure if it's worth fixing that.

File: torch/_subclasses/fake_tensor.py (hunk 12 above)
Comment (peterbell10) [reply torch/_subclasses/fake_tensor.py:169]: Makes sense, can we do something more like `decomposition_table[func].__module__.startswith("torch._decomp")`? This would prevent picking up functions with the same name but defined in a different place.

**Review by peterbell10 — APPROVED** [review]

**Comment by lezcano** [comment]
rebasing on main as viable/strict has been broken for quite some time

File: torch/_inductor/codegen/cpp.py [hunk 15]
```diff
@@ -804,6 +805,23 @@ def atanh(x):
     def copysign(x, y):
         return f"std::copysign({x}, {y})"
 
+    @staticmethod
+    def frexp(x):
+        cache_keys = f"frexp({x})[0]", f"frexp({x})[1]"
+        if all(cache_key in V.kernel.cse.cache for cache_key in cache_keys):
+            return tuple(V.kernel.cse.cache[cache_key] for cache_key in cache_keys)
```
Comment (peterbell10) [inline torch/_inductor/codegen/cpp.py:812]: Why was this change needed?

Comment (isuruf) [reply torch/_inductor/codegen/cpp.py:812]: https://github.com/pytorch/pytorch/blob/d9db9e62e3d2d58d4e76a43f30c15db389e51c17/torch/_inductor/codegen/cpp.py#L1561 assumes that the cache values are variables and not tuples. I can try to change that function to not assume that if that's preferred.

Comment (peterbell10) [reply torch/_inductor/codegen/cpp.py:812]: Hrm okay, a bit less elegant but oh well

**Comment by isuruf** [comment]
Failing test is due to https://github.com/pytorch/pytorch/issues/119364

**Comment by clee2000** [comment]
I'm pretty sure something in this stack broke test_unary_ufuncs.py::TestUnaryUfuncsCUDA::test_reference_numerics_extremal__refs_frexp_cuda_float64 (and  other variants of it) on windows.  @isuruf  Could you revert or provide a forward fix?
example logs https://hud.pytorch.org/pytorch/pytorch/commit/79f059987eead42b9279e31e28ed6fd697630d36
https://github.com/pytorch/pytorch/actions/runs/8028300864/job/21933571465
https://github.com/pytorch/pytorch/actions/runs/8028300864/job/21933571716

**Comment by isuruf** [comment]
exponent for `inf` and `NaN` are undefined and the `_refs.frexp` function added in this PR added a new test. Let's mark the new test as xfail.
